### ISSUE TITLE
Blazor localization not working

### ISSUE THREAD

**Issue by Stamo-Gochev** [dotnet/aspnetcore#16687]
### Describe the bug
I am trying to localize a custom component (the component itself, not the page it is displayed in), but the `.resx` files are not being used at all.

### To Reproduce
1. Clone the repo at: https://github.com/Stamo-Gochev/blazor-test-localization
2. Try to change the culture using the dropdown, e.g. from "English" to "Francais".

### Expected behavior
The `LocalizedComponent` text is updated according to the culture of the app.

### Additional context

The [LocalizedComponent](https://github.com/Stamo-Gochev/blazor-test-localization/blob/164e4b4a32b2dd428993ee68f26b721da469490e/blazor-test-localization/Components/LocalizedComponent.razor#L8) uses [IStringLocalizer](https://github.com/Stamo-Gochev/blazor-test-localization/blob/164e4b4a32b2dd428993ee68f26b721da469490e/blazor-test-localization/Components/LocalizedComponent.razor.cs#L16). A `select` element is added (similar to the docs) in order to change the [culture](https://github.com/Stamo-Gochev/blazor-test-localization/blob/164e4b4a32b2dd428993ee68f26b721da469490e/blazor-test-localization/Components/LocalizedComponent.razor#L34-L41).

Changing the culture to `fr` using the dropdown changes the format of the [DateTime](https://github.com/Stamo-Gochev/blazor-test-localization/blob/164e4b4a32b2dd428993ee68f26b721da469490e/blazor-test-localization/Components/LocalizedComponent.razor#L11), but the [localized value](https://github.com/Stamo-Gochev/blazor-test-localization/tree/master/blazor-test-localization/Resources) remains the same. Actually, it displays the key - `key1`, so I suppose that the localization is not triggered at all. Resources are registered in [Startup.cs](https://github.com/Stamo-Gochev/blazor-test-localization/blob/164e4b4a32b2dd428993ee68f26b721da469490e/blazor-test-localization/Startup.cs#L33). Can you shed some light why this isn't working?

**Comment by campersau** [comment]
Using resource files will create additional satellite assemblies which aren't downloaded by blazor.
But you can embedd them in the main assembly. See https://github.com/aspnet/AspNetCore/issues/15270

**Comment by javiercn** [comment]
@Stamo-Gochev thanks for contacting us.

Can you provide a minimal repro for your issue? 

We have end to end automation for this area, so I suspect there must be a mismatch in the docs or in what you are doing.

@rynowak is probably the right person to look at this.

**Comment by Stamo-Gochev** [comment]
@javiercn Yes, the link to the repo is listed in the steps to reproduce. Here it is:
https://github.com/Stamo-Gochev/blazor-test-localization

**Comment by javiercn** [comment]
@Stamo-Gochev Thanks, I didn't see it the first time.

**Comment by Stamo-Gochev** [comment]
@mkArtakMSFT The issue I referenced mentions that limited functionality is supported, including `IStringLocalizer` https://github.com/aspnet/AspNetCore.Docs/issues/13436:
>A limited set of ASP.NET Core's localization features are supported in Blazor at the time of writing. **IStringLocalizer<> can be used with Blazor**. IHtmlLocalizer<>, IViewLocalizer<> and Data Annotations localization are ASP.NET Core MVC features and cannot be used with Blazor. https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-2.2#make-the-apps-content-localizable

This is part of the docs at: https://docs.microsoft.com/en-us/aspnet/core/blazor/components?view=aspnetcore-3.0#use-net-localization-scenarios-in-blazor-apps
> A limited set of ASP.NET Core's localization scenarios are currently supported:
IStringLocalizer<> is supported in Blazor apps.
IHtmlLocalizer<>, IViewLocalizer<>, and Data Annotations localization are ASP.NET Core MVC scenarios and not supported in Blazor apps.

This is why I thought that it is supported in server-side blazor. Is there an alternative that is currently supported?

**Comment by rynowak** [comment]
@Stamo-Gochev - this line is commented out https://github.com/Stamo-Gochev/blazor-test-localization/blob/master/blazor-test-localization/Components/LocalizedComponent.razor#L43

You **need** to do the redirect for this to work in server-side Blazor. Updating the thread culture directly won't cause the UI to update.

If you want an alternative to `IStringLocalizer` you can you use the generated class that corresponds to your resource file. Normally when you add a `.resx` file from Visual Studio it will also add a generated class that provides access to all of the strings.

**Comment by Stamo-Gochev** [comment]
@rynowak I added the redirection in https://github.com/Stamo-Gochev/blazor-test-localization/commit/6de7aba7a80354575486d8514fbb2d1bee25cd0d but I am still not able to make the localization work even though the culture controller action is hit. Is there anything wrong with the update?

**Comment by Stamo-Gochev** [comment]
@rynowak Can you at least confirm if `StringLocalizer` is supported for localizing blazor components or not?

**Comment by rynowak** [comment]
I'll try your sample out and see what's going on.

**Comment by Stamo-Gochev** [comment]
Are there any updates on that? Is there anything else that I can do in order to help with the investigation?

**Comment by rynowak** [comment]
@Stamo-Gochev - sorry for the delay. The problem that you're running into is a flaw in the localization system. It doesn't work well when the assembly name is different from the project name. 

I took your sample and change the assembly name to match the namespace, and it started working. In general this is a problem when the project name contains characters that are not allowed in a C# identifier - so `blazor-test-localization` gets transformed into `blazor_test_localization`.

I've logged https://github.com/aspnet/AspNetCore.Docs/issues/15673 to track improving the documentation.
